verification/article/article.md

```diff
@@ -0,0 +1,1063 @@
+
+### 4.1. Asset Conversion (`AssetCalculations`)
+
+This module defines the logic for converting a base `amount` into assets based on a price fraction
+`saleToken / depositToken`. The specification is:
+`CalculateAssetsSpec(amount, dT, sT) := (amount * sT) / dT`
+
+The module provides lemmas that instantiate the generic mathematical properties for this specific context. For instance,
+`Lemma_CalculateAssets_IsGreaterOrEqual` proves that `CalculateAssetsSpec(...) >= amount` if `sT >= dT`, by directly
+invoking `Lemma_MulDivGreater_From_Scratch`.
+
+A critical property for refund safety is the **round-trip inequality**, proven in
+`Lemma_AssetsRevert_RoundTrip_lte` [@bhargavan2016formal]. It
+states that converting an amount to assets and then immediately converting those assets back to the original currency
+cannot result in a gain.
+
+* **Property 4 (Asset Conversion Round-Trip Safety).** Let $Assets(w) := \text{CalculateAssetsSpec}(w, dT, sT)$ and
+  $Revert(a) := \text{CalculateAssetsRevertSpec}(a, dT, sT)$. Then for $w > 0$:
+  $$ \text{Assets}(w) > 0 \implies \text{Revert}(\text{Assets}(w)) \le w $$
+
+### 4.2. Time-Based Discounts (`Discounts`)
+
+This module implements percentage-based bonuses. It uses fixed-point arithmetic with a `MULTIPLIER` of 10000 to
+represent percentages with four decimal places. It also verifies a critical business rule: discount periods must not
+overlap.
+
+* **Property 5 (Discount Non-Overlap).** For a sequence of discounts $D$, the following predicate holds:
+  $$ \forall i, j. (0 \le i < j < |D|) \implies (D_i.\text{endDate} \le D_j.\text{startDate} \lor D_j.\text{endDate} \le D_i.\text{startDate}) $$
+
+Dafny successfully proves that this property implies the uniqueness of any active discount at a given time (
+`Lemma_UniqueActiveDiscount`), which is essential for ensuring that deposit calculations are deterministic and
+unambiguous [@grishchenko2018semantic]. Similar to asset conversions, the module also proves the round-trip safety for
+applying and reverting a
+discount (`Lemma_WeightOriginal_RoundTrip_lte`).
+
+## 5. Top-Level Specification and State Machine Verification
+
+The verified components are composed at the top layers to model the complete system behavior.
+
+### 5.1. The Deposit Workflow and Refund Safety (`Deposit` module)
+
+This module specifies the end-to-end logic for a user deposit. The main function, `DepositSpec`, branches based on the
+sale mechanic. The most complex case is `DepositFixedPriceSpec`, which handles deposits into a sale with a hard cap (
+`saleAmount`). If a deposit would cause the total sold tokens to exceed this cap, a partial refund must be calculated.
+
+The paramount safety property for this entire system is ensuring that this refund never exceeds the user's initial
+deposit. This is formally stated and proven in `Lemma_RefundIsSafe`.
+
+* **Property 6 (Ultimate Refund Safety).** For a valid configuration and any deposit `amount`, the calculated `refund`
```

Comment:
Possible future work:

Continuing with the narrative of users not losing funds, I think a better property to show would be `refund + user_balance = amount` where `user_balance` is the funds the contract currently has allocated to the user. Note that since `0 <= user_balance`, this property has `refund <= amount` as a corollary.

Reply:
Resolved in #76